### ISSUE TITLE
Consider using fasthttp instead of net/http

### ISSUE THREAD

**Issue by oxtoacart** [getlantern/lantern#3790]
It's faster :) https://github.com/valyala/fasthttp/

It's not a drop-in replacement though, so need to consider whether it's worth the effort.

cc: @myleshorton

**Comment by uaalto** [comment]
We have also talked about this for the proxy: https://github.com/getlantern/http-proxy/issues/68

**Comment by aranhoide** [comment]
@oxtoacart feel free to bump priority/milestone if you think `net/http` is a bottleneck likely to affect users' experience.

**Comment by oxtoacart** [comment]
@aranhoide I think the only thing that might make this higher priority is if it solves our issue with not being able to upgrade http-proxy to Go 1.6.

**Comment by aranhoide** [comment]
@oxtoacart interesting.  Do we have an issue for that?  Maybe adding a reference from that one to this one would be a good way to keep track of this conditional priority.

**Comment by oxtoacart** [comment]
Switching to fasthttp may resolve getlantern/http-proxy-lantern#37.  Not sure though - would need to test.

**Comment by uaalto** [comment]
It might be a dangerous path. We tested extensively the current server, and is supported by the Go team. The promise of faster server comes with drawbacks, and AFAIK about how it handles edge cases. For a proxy, it makes sense to be more standard, while fasthttp seems like a good option if you control all the calls performed to the server, so you don't care about the full HTTP specification. BTW, not even the Go server supports the full specification.

**Comment by oxtoacart** [comment]
@uaalto Good point.

**Comment by oxtoacart** [comment]
@uaalto Question - why do we need our proxy to be a fully HTTP compliant proxy, or even close to it?  It's only our client that communicates with it.  And to a certain extent, it can be a dumb proxy because we don't care about caching and such things.

**Comment by uaalto** [comment]
Yes, but our client is not generating the requests, so it doesn't matter that only the Lantern client uses it, we need to handle whatever the browser passes to it. This only applies for forwarding of course, so everything going through HTTP connect is much simpler.
In order to see some of the requests that a browser could send to our client that need to be properly parsed and forwarded to the proxies, we've used Co Advisor. One example of such thing that is not completely handled at the moment is Trailers. Incidentaly, most browsers don't support them, which is the main limitation to use gRPC-on-the-browser, but we need to take care of as many of those as possible, because requests that come through are not generated by us. And, most of this is handled by the Go server, we just tweak a few higher-level things.

**Comment by oxtoacart** [comment]
Interesting.  In theory, we could treat plain HTTP similarly to connect traffic.

1. Hijack the connection on the client
2. Send the raw data to the server
3. Wait for the Host: header to determine where to connect to
4. Make the connection to the upstream server
5. Shuttle the traffic across just like with a CONNECT

So in essence, the remote proxies wouldn't really be HTTP proxies anymore, they'd be more like SOCKS proxies.

**Comment by oxtoacart** [comment]
Of course they would still pretend to be HTTP proxies for the bad guys :)

**Comment by uaalto** [comment]
We do shuttle the body (well, trailers are part of the body sort of, but let's leave them aside because they are not seen often AFAIK). But the headers need some kind of processing generally (see forward.go). So, for every single request you receive you have to perform the same dance. In the case of CONNECT you just do it once, and it is much simpler.
So, we would still need to handle headers and do our thing with them, then send the body and receive the response. Which is what we already do. If we don't do any of these, it could be considered closer to CONNECT's behavior, but the thing is that we need to do it in each request.
